### DISCUSSION TITLE
How do I call a method on an 'auto disposing' Notifier before the first watcher registers?

### DISCUSSION THREAD

**Discussion by wujek-srujek** [rrousselGit/riverpod#3182] [Q&A]
I have an `AutoDisposeNotifier` and a corresponding `NotifierProvider.autoDispose<...>`. The notifier has a method, let's call it `start`, that triggers execution of the logic.

In my use case, the users clicks a button and another route is pushed, with a widget that `watch`es the provider. I would like the `start` method to be called automatically, before this widget is first drawn, and I have some kind of a mental block and I don't know how to do it. I tried various things:

1. Called `ref.read(provider.notifier).start()` right before navigation - this however initializes the notifier, calls `start`, and then immediately disposes it. The reason is that there are no listeners and the provider is 'auto disposing' (and I need it to stay this way).
2. Made the target route widget a `ConsumerStatefulWidget` and tried calling `ref.read(provider.notifier).start()` in `initState` - this fails an assertion that I'm changing the notifier in a widget lifecycle.
3. Then I used the `WidgetsBinding.instance.addPostFrameCallback(...)` in `initState`, and called `ref.read(provider.notifier).start()` in the callback - this actually works, but the UI flickers initially because the widget is first drawn with the initial state (before `start`) and then almost immediately is redrawn with the new state (after `start`). The same happens when I use a `Future(() => ...)` in `initState`. This is not the UX I want.

All works fine apart from that, when I trigger `start` using a button or something like this. But how do I trigger it without a button?

**Comment by Colman** [comment]
I've ran into a similar problem. The way I solved it was by doing this where you make the navigation:

```dart
final subscription = ref.listenManual(provider, (_, __) {});

try {
   ref.read(provider.notifier).start();
   Navigator.push(context, ...);
} finally {
   WidgetsBinding.instance.addPostFrameCallback((_) {
      subscription.close();
   });
}
```

It will create the provider instance when you call `listenManual` and will keep it alive at least until one frame after you navigate. This way it's guaranteed to still be alive on the first frame of the page you are navigating to. Then you close the subscription in the post frame callback. This way, after the first frame, the only thing that will be listening to it is the watcher in the page you navigate to. So if you leave that page then it will be disposed correctly.

Also, if your navigation redirects you for some reason, then the subscription will close after the next frame which will dispose the provider. So worst case, your provider takes up memory for an extra frame.
